client/web/src/repo/RepoContainer.tsx

```
@@ -272,24 +272,31 @@ export const RepoContainer: React.FunctionComponent<RepoContainerProps> = props
         }, [repoOrError, resolvedRevisionOrError, props.history, props.location])
     )
 
-    // Update the workspace roots service to reflect the current repo / resolved revision
+    // Update the workspace roots via the extension host API to reflect the current repo / resolved revision
     useEffect(() => {
-        props.extensionsController.services.workspace.roots.next(
+        const workspaceRootURI =
             resolvedRevisionOrError && !isErrorLike(resolvedRevisionOrError)
-                ? [
-                      {
-                          uri: makeRepoURI({
-                              repoName,
-                              revision: resolvedRevisionOrError.commitID,
-                          }),
-                          inputRevision: revision || '',
-                      },
-                  ]
-                : []
-        )
-        // Clear the Sourcegraph extensions model's roots when navigating away.
-        return () => props.extensionsController.services.workspace.roots.next([])
-    }, [props.extensionsController.services.workspace.roots, repoName, resolvedRevisionOrError, revision])
+                ? makeRepoURI({ repoName, revision: resolvedRevisionOrError.commitID })
+                : undefined
+        if (workspaceRootURI) {
+            props.extensionsController.extHostAPI
+                .then(extensionHostAPI =>
+                    extensionHostAPI.addWorkspaceRoot({
```

The old logic is: if there's a root then emit it, but otherwise emit an empty array which clears the roots. In the new logic is it sufficient to add a root if there is one, but only remove it below in the useEffect's unmount function?

---

The useEffect function isn't just an unmount function, right? It gets called whenever the repo or revision changed. Only thing that may be an issue is that we're not waiting with adding the new root after the old is removed (which may not be a real problem) and that we're swallowing errors (we'll just continue interacting with the extension host despite the root not being added).